File: framework/dotnet/BetterPortal/Clients.cs
using AnyVali;
using System.Text;
using System.Text.Json;
using Node = System.Collections.Generic.Dictionary<string, object?>;

namespace BetterPortal;

public sealed class ClientException(int status, string message) : Exception(message)
{
    public int Status { get; } = status;
}

public sealed record ClientOperationSchemas(IReadOnlyDictionary<string, Node> Inputs, IReadOnlySet<string> RequiredInputs, Node Output);

/// <summary>A dependency's exported BP contract. Inputs and outputs use its AnyVali documents.</summary>
public sealed class ClientContract
{
    internal Node Document { get; }
    public string PluginId => (string)((Node)Document["manifest"]!)["pluginId"]!;
    private sealed record Entry(Node View, Node Operation, Schema Keys, Dictionary<string, Schema> Fields, Schema? Output);
    private readonly Dictionary<string, Entry> operations = new(StringComparer.Ordinal);
    public ClientContract(object value)
    {
        Document = (Node)Contracts.Parse("BpSchemaOutputSchema", value)!;
        foreach (var view in Items((Node)Document["manifest"]!, "views"))
        {
            foreach (var path in new[] { (string)view["path"]! }.Concat(((List<object?>)view["pathVariants"]!).Cast<string>())) _ = Route.Segments(path);
            foreach (var operation in Items(view, "operations"))
            {
                var fields = new Node { ["params"] = view["paramsSchema"], ["query"] = operation["querySchema"], ["headers"] = operation["headersSchema"] };
                var body = (Node)operation["bodySchema"]!;
                if (body.Count == 0)
                {
                    body = Contracts.Document("JsonObjectSchema");
                    body["root"] = new Node { ["kind"] = "optional", ["inner"] = body["root"] };
                }
                fields["body"] = body;
                var schemas = new Dictionary<string, Schema>(); var keys = new Dictionary<string, Node>();
                foreach (var (name, field) in fields)
                {
                    var document = (Node)Json.Read(Json.Write(field is Node { Count: > 0 } ? field : Contracts.Document("JsonObjectSchema")))!;
                    if (!document.ContainsKey("root")) throw new ArgumentException("Dependency schemas must be AnyVali documents");
                    if (name != "body") ((Node)document["root"]!).TryAdd("default", new Node());
                    // Separate wrappers preserve independent definitions and omitted-versus-null fields.
                    schemas[name] = Contracts.Import(Contracts.ObjectDocument(new Dictionary<string, Node> { [name] = document }, "reject"));
                    var optional = Contracts.Document("JsonValueSchema");
                    optional["root"] = new Node { ["kind"] = "optional", ["inner"] = optional["root"] }; keys[name] = optional;
                }
                var entry = new Entry(view, operation, Contracts.Import(Contracts.ObjectDocument(keys, "reject")), schemas,
                    operation["jsonResponseSchema"] is Node { Count: > 0 } response ? Contracts.Import(response) : null);
                if (!operations.TryAdd((string)operation["operationId"]!, entry)) throw new ArgumentException("Duplicate dependency operation ID");
            }
        }
    }
    internal static IEnumerable<Node> Items(Node node, string key) => ((List<object?>)node.GetValueOrDefault(key, new List<object?>())!).Cast<Node>();
    public Generated.BpSchemaOutput Schema() => Contracts.Parse<Generated.BpSchemaOutput>("BpSchemaOutputSchema", Document);
    /// <summary>Owned documents for native authoring. AnyVali determines input presence.</summary>
    public IReadOnlyDictionary<string, ClientOperationSchemas> JsonOperations()
    {
        var result = new Dictionary<string, ClientOperationSchemas>(StringComparer.Ordinal);
        foreach (var (identifier, entry) in operations)
        {
            if (entry.Output is null) continue;
            var inputs = new Dictionary<string, Node>();
            foreach (var (name, field) in entry.Fields)
            {
                var document = (Node)Json.Read(Json.Write(V.Export(field)))!;
                document["root"] = ((Node)((Node)document["root"]!)["properties"]!)[name]; inputs[name] = document;
            }
            result[identifier] = new(inputs, entry.Fields.Where(pair => !pair.Value.SafeParse(new Node()).Success).Select(pair => pair.Key).ToHashSet(StringComparer.Ordinal),
                (Node)Json.Read(Json.Write(V.Export(entry.Output)))!);
        }
        return result;
    }
    internal (Node View, Node Operation, Node Values, string Path, string Variant, Schema Output) Prepare(string identifier, object? values)
    {
        if (!operations.TryGetValue(identifier, out var entry)) throw new ArgumentException("Unknown dependency operation");
        if (entry.Output is null) throw new ArgumentException("Dependency operation does not declare JSON output");
        var source = (Node)Contracts.Parse(entry.Keys, values ?? new Node())!; var parsed = new Node();
        foreach (var (name, schema) in entry.Fields)
            foreach (var pair in (Node)Contracts.Parse(schema, source.TryGetValue(name, out var value) ? new Node { [name] = value } : new Node())!) parsed.Add(pair.Key, pair.Value);
        if (new[] { "params", "query", "headers" }.Any(name => parsed.GetValueOrDefault(name) is not Node))
            throw new ArgumentException("Dependency params, query and headers must be objects");
        var selected = new[] { (string)entry.View["path"]! }.Concat(((List<object?>)entry.View["pathVariants"]!).Cast<string>())
            .Select(variant => (Variant: variant, Path: Urls.Fill(variant, (Node)parsed["params"]!))).FirstOrDefault(value => value.Path is not null);
        if (selected.Path is null) throw new ArgumentException("Dependency route parameters do not select a path");
        return (entry.View, entry.Operation, parsed, selected.Path, selected.Variant, entry.Output);
    }
}

/// <summary>One connection pool per service; BP owns credentials, destinations and limits.</summary>
public sealed class ServiceClients : IAsyncDisposable
{
    internal Service Service { get; }
    private readonly Lazy<HttpClient> http = new(() => new(new SocketsHttpHandler { AllowAutoRedirect = false, UseProxy = false, UseCookies = false }) { Timeout = Timeout.InfiniteTimeSpan });
    private readonly object gate = new();
    private readonly HashSet<TaskCompletionSource> pending = [];
    private readonly CancellationTokenSource shutdown = new();
    private bool closed;
    internal ServiceClients(Service service) => Service = service;
    /// <summary>Background callers may use declared service-mode dependencies only.</summary>
    public RequestClients Scope(string tenantId, string appId) => new(this, tenantId, appId);
    internal async Task<object?> Send(Client client, string identifier, object? values, CancellationToken cancellation)
    {
        var done = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
        lock (gate) { if (closed) throw new ClientException(503, "Dependency client is closed"); pending.Add(done); }
        using var timeout = CancellationTokenSource.CreateLinkedTokenSource(cancellation, Service.Stopping, client.Context.Cancellation, shutdown.Token);
        timeout.CancelAfter(TimeSpan.FromSeconds(30));
        try
        {
            var (view, operation, parsed, path, variant, output) = client.Contract.Prepare(identifier, values);
            var (state, scope) = client.Context.Current();
            var (origin, headers) = client.Authorize(state, scope, view, operation, variant);
            var custom = ((Node)parsed["headers"]!).Select(pair => new KeyValuePair<string, string>(pair.Key,
                pair.Value is string or bool or sbyte or byte or short or ushort or int or uint or long or ulong or float or double or decimal
                    ? Urls.Scalar(pair.Value) : throw new ArgumentException("Dependency header values must be strings, numbers or booleans"))).ToArray();
            _ = new RawResponse(headers: custom);
            string[] forbidden = ["authorization", "cookie", "set-cookie", "host", "origin", "referer", "accept", "accept-encoding", "content-type", "content-encoding"];
            if (custom.Select(pair => pair.Key).Distinct(StringComparer.OrdinalIgnoreCase).Count() != custom.Length
                || custom.Any(pair => forbidden.Contains(pair.Key, StringComparer.OrdinalIgnoreCase) || pair.Key.StartsWith("x-bp-", StringComparison.OrdinalIgnoreCase)))
                throw new ArgumentException("Dependency headers cannot replace framework credentials or routing");
            foreach (var (name, value) in custom) headers[name] = value;
            var pairs = new List<KeyValuePair<string, string>>();
            foreach (var (name, value) in (Node)parsed["query"]!)
                foreach (var item in value is List<object?> array ? array : [value])
                    if (item is not null) pairs.Add(new(name, Urls.Scalar(item)));
            using var query = new FormUrlEncodedContent(pairs);
            var address = origin + path + (pairs.Count > 0 ? "?" + await query.ReadAsStringAsync(timeout.Token) : "");
            var data = parsed.TryGetValue("body", out var body) ? Encoding.UTF8.GetBytes(Json.Write(body)) : null;
            if (address.Length > 8192 || data?.Length > 16 * 1024 * 1024) throw new ArgumentException("Dependency request exceeds its size limit");
            var method = (string)operation["method"]!;
            if (data is not null && method is "GET" or "HEAD") throw new ArgumentException("Dependency GET/HEAD requests cannot carry a body");
            using var request = new HttpRequestMessage(new HttpMethod(method), address);
            if (data is not null) { request.Content = new ByteArrayContent(data); request.Content.Headers.ContentType = new("application/json"); }
            headers["accept"] = "application/json";
            headers["accept-encoding"] = "identity";
            foreach (var (name, value) in headers)
                if (!request.Headers.TryAddWithoutValidation(name, value)) throw new ArgumentException("Unsupported dependency request header");
            try
            {
                using var response = await http.Value.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, timeout.Token);
                var status = (int)response.StatusCode;
                if (!response.IsSuccessStatusCode) throw new ClientException(status, "Dependency returned an unsuccessful status");
                if (status is not (204 or 205) && response.Content.Headers.ContentType?.MediaType?.ToLowerInvariant() != "application/json")
                    throw new ClientException(502, "Dependency did not return JSON");
                if (response.Content.Headers.ContentEncoding.Any(value => !value.Equals("identity", StringComparison.OrdinalIgnoreCase)))
                    throw new ClientException(502, "Dependency returned unsupported content encoding");
                await using var stream = await response.Content.ReadAsStreamAsync(timeout.Token);
                using var buffer = new MemoryStream(); var chunk = new byte[8192]; int count;
                while ((count = await stream.ReadAsync(chunk, timeout.Token)) != 0)
                {
                    if (buffer.Length + count > 16 * 1024 * 1024) throw new ClientException(502, "Dependency response exceeds its size limit");
                    buffer.Write(chunk, 0, count);
                }
                client.Context.Current(state);
                return Contracts.Parse(output, status is 204 or 205 && buffer.Length == 0 ? null : Json.Read(new UTF8Encoding(false, true).GetString(buffer.ToArray())));
            }
            catch (ClientException) { throw; }
            catch (OperationCanceledException) when (cancellation.IsCancellationRequested || Service.Stopping.IsCancellationRequested || client.Context.Cancellation.IsCancellationRequested || shutdown.IsCancellationRequested) { throw; }
            catch (Exception) { throw new ClientException(502, "Dependency request or response validation failed"); }
        }
        finally { lock (gate) pending.Remove(done); done.TrySetResult(); }
    }
    public async ValueTask DisposeAsync()
    {
        Task[] tasks; lock (gate) { if (closed) return; closed = true; tasks = pending.Select(item => item.Task).ToArray(); }
        await shutdown.CancelAsync();
        await Task.WhenAll(tasks);
        if (http.IsValueCreated) http.Value.Dispose();
    }
}

/// <summary>Opaque request binding. User credentials never enter render contexts.</summary>
public sealed class RequestClients
{
    internal ServiceClients Owner { get; }
    private readonly string tenant, app;
    private readonly Service.SnapshotState? revision;
    internal string? UserToken { get; }
    internal CancellationToken Cancellation { get; }
    internal RequestClients(ServiceClients owner, string tenant, string app, Service.SnapshotState? revision = null, string? userToken = null, CancellationToken cancellation = default)
    { Owner = owner; this.tenant = tenant; this.app = app; this.revision = revision; UserToken = userToken; Cancellation = cancellation; }
    internal (Service.SnapshotState State, ScopedContext Scope) Current(Service.SnapshotState? expected = null) => Owner.Service.ClientState(tenant, app, revision, expected);
    public Client User(ClientContract contract, string? serviceId = null) => revision is null
        ? throw new ArgumentException("User clients require a service request context") : new(this, contract, serviceId ?? contract.PluginId);
    public Client M2m(string requestId, ClientContract contract) => new(this, contract, requestId: requestId);
}

public sealed class Client
{
    internal RequestClients Context { get; }
    internal ClientContract Contract { get; }
    private readonly string? serviceId, requestId;
    internal Client(RequestClients context, ClientContract contract, string? serviceId = null, string? requestId = null)
    { Context = context; Contract = contract; this.serviceId = serviceId; this.requestId = requestId; }
    public Task<object?> RequestAsync(string operationId, object? values = null, CancellationToken cancellation = default) => Context.Owner.Send(this, operationId, values, cancellation);
    public async Task<T> RequestAsync<T>(string operationId, object? values = null, CancellationToken cancellation = default)
        => JsonSerializer.Deserialize<T>(Json.Write(await RequestAsync(operationId, values, cancellation)), Json.Options)!;
    private static IEnumerable<Node> Items(Node node, string key) => ClientContract.Items(node, key);
    internal (string Origin, Dictionary<string, string> Headers) Authorize(Service.SnapshotState state, ScopedContext scope, Node view, Node operation, string path)
    {
        var service = Context.Owner.Service; var auth = (Node)operation["auth"]!;
        var callers = (List<object?>)auth["callers"]!; var method = (string)operation["method"]!;
        Node target; var headers = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        if (requestId is null)
        {
            var reference = service.Registry.Dependencies.GetValueOrDefault(serviceId!, serviceId!);
            var targets = Items(scope.Tenant, "services").Where(item => item["enabled"] is true && (Equals(item["id"], reference) || Equals(item.GetValueOrDefault("serviceId"), reference)) && Equals(item.GetValueOrDefault("serviceId"), Contract.PluginId)).ToArray();
            if (targets.Length != 1) throw new ClientException(403, "Dependency service is unavailable or ambiguous");
            target = targets[0];
            if (!callers.Contains("user")) throw new ClientException(403, "Dependency operation does not allow user calls");
            var app = new Node(scope.App) { ["routes"] = scope.App.GetValueOrDefault("appRoutes", scope.App["routes"]) };
            var access = new AppAccess(new(scope.Tenant, app, scope.OriginPolicy), [(string)target["id"]!]);
            if (!access.AllowsOperation((string)view["viewId"]!, (string)operation["operationId"]!, method, path, serviceId: (string)target["id"]!)) throw new ClientException(403, "Dependency operation is not mounted");
            headers["origin"] = ((List<object?>)scope.App["hostnames"]!).Cast<string>().SelectMany(HttpAddress.Origins).FirstOrDefault()
                ?? throw new ClientException(503, "Dependency app origin is unavailable");
            if (Context.UserToken is { } user) headers["authorization"] = "Bearer " + user;
            else if (auth["required"] is true) throw new ClientException(401, "Dependency requires a BP user");
        }
        else
        {
            var manifest = (Node)Contracts.Parse("PluginManifestSchema", service.Manifest)!;
            var requests = Items(manifest, "m2mRequests").Where(item => Equals(item["id"], requestId)).ToArray();
            if (requests.Length != 1) throw new ClientException(403, "Dependency request is not uniquely declared");
            var request = requests[0]; var mode = (string)request["mode"]!;
            if (!callers.Contains(mode) || !((List<object?>)request["methods"]!).Contains(method)) throw new ClientException(403, "Dependency caller mode or method is not allowed");
            var policy = (Node)state.Config.GetValueOrDefault("m2m", new Node())!;
            var bindings = Items(policy, "bindings").Where(item => item["enabled"] is true && Equals(item["requestId"], requestId)
                && ((List<object?>)policy["localServiceIds"]!).Contains(item["sourceServiceId"]) && Equals(item["tenantId"], scope.TenantId)
                && Equals(item.GetValueOrDefault("appId", scope.AppId), scope.AppId) && Equals(item["mode"], mode)
                && Equals(item["contractId"], request["contractId"]) && Equals(item["targetViewId"], view["viewId"])).ToArray();
            if (bindings.Length != 1) throw new ClientException(403, "Dependency binding is unavailable or ambiguous");
            var binding = bindings[0];
            var contracts = Items((Node)Contract.Document["manifest"]!, "apiContracts").Where(item => Equals(item["id"], binding["contractId"]) && Equals(item["viewId"], view["viewId"])).ToArray();
            if (contracts.Length != 1 || !Items(operation, "apiContracts").Any(item => Equals(item["id"], binding["contractId"])) || !((List<object?>)contracts[0]["methods"]!).Contains(method) || !((List<object?>)contracts[0]["modes"]!).Contains(mode))
                throw new ClientException(403, "Dependency contract does not cover the operation");
            if (!((List<object?>)request["requiredCapabilities"]!).ToHashSet().IsSubsetOf((List<object?>)contracts[0]["capabilities"]!))
                throw new ClientException(403, "Dependency contract capabilities are insufficient");
            var required = ((List<object?>)request["permissions"]!).Concat((List<object?>)contracts[0]["permissions"]!)
                .Concat(Items(auth, "permissions").SelectMany(item => (List<object?>)item["permissions"]!)).ToHashSet();
            if (!Items(policy, "grants").Any(item => item["enabled"] is true && Equals(item["bindingId"], binding["id"]) && Equals(item["tenantId"], scope.TenantId)
                && Equals(item.GetValueOrDefault("appId", scope.AppId), scope.AppId) && ((List<object?>)item["methods"]!).Contains(method) && required.IsSubsetOf((List<object?>)item["permissions"]!)))
                throw new ClientException(403, "Dependency grant is unavailable or insufficient");
            var targets = Items(policy, "services").Where(item => Equals(item["id"], binding["targetServiceId"]) && Equals(item.GetValueOrDefault("serviceId"), Contract.PluginId)).ToArray();
            if (targets.Length != 1) throw new ClientException(403, "Dependency target is unavailable or ambiguous");
            target = targets[0]; var key = service.SigningKey; var identity = (Node)state.Config.GetValueOrDefault("serviceIdentity", new Node())!;
            static string Pem(string value) => string.Concat(value.Where(c => !char.IsWhiteSpace(c)));
            if (key is null || !Equals(identity.GetValueOrDefault("keyId"), key.Kid) || Pem((string)identity.GetValueOrDefault("publicKeyPem", "")!) != Pem(key.PublicKeyPem))
                throw new ClientException(503, "Dependency signing identity is not registered");
            headers["x-bp-service-id"] = (string)binding["sourceServiceId"]!; headers["x-bp-tenant-id"] = scope.TenantId; headers["x-bp-app-id"] = scope.AppId;
            var now = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            var token = Tokens.Sign(key, new Node { ["iss"] = binding["sourceServiceId"], ["sub"] = binding["sourceServiceId"], ["aud"] = target["id"],
                ["tenantId"] = scope.TenantId, ["appId"] = scope.AppId, ["bindingId"] = binding["id"], ["iat"] = now, ["exp"] = now + 60, ["jti"] = Guid.CreateVersion7().ToString(), ["tokenType"] = "service" }, TokenPurpose.Service);
            if (mode == "delegated")
            {
                if (Context.UserToken is not { } user) throw new ClientException(401, "Delegated dependency calls require a BP user");
                headers["authorization"] = "Bearer " + user; headers["x-bp-service-authorization"] = "Bearer " + token;
            }
            else headers["authorization"] = "Bearer " + token;
        }
        return (HttpAddress.Origin(TrustedKeys.SecureEndpoint((string)target["hostname"]!).AbsoluteUri), headers);
    }
}

public sealed partial class Service
{
    internal KeyPair? SigningKey { get; set; }
    private readonly Lazy<ServiceClients> clients;
    public ServiceClients Clients => clients.Value;
    internal (SnapshotState State, ScopedContext Scope) ClientState(string tenant, string app, SnapshotState? revision, SnapshotState? expected)
    {
        var current = state;
        if (!Ready || current is null || revision is not null && !ReferenceEquals(current, revision) || expected is not null && !ReferenceEquals(current, expected))
            throw new ClientException(503, "Dependency configuration is unavailable or changed");
        if (installedTenantLock is not null && tenant != installedTenantLock) throw new ClientException(403, "Dependency tenant is not allowed");
        return (current, current.Snapshot.ById(tenant, app) ?? throw new ClientException(403, "Dependency tenant/app is unavailable"));
    }
}

File: framework/python/betterportal/clients.py
"""Scoped JSON dependency clients. BP owns credentials, policy and transport limits."""
from __future__ import annotations

import asyncio
from copy import deepcopy
from dataclasses import dataclass
from http.cookiejar import CookieJar, DefaultCookiePolicy
import json
import time
from typing import Any, TYPE_CHECKING
from urllib.parse import urlencode

import anyvali as av
import httpx

from .access import AppAccess
from .context import ScopedContext, http_origin, _origins
from .contracts import document, object_document, parse
from .jsoncodec import loads
from .keys import secure_endpoint
from .response import RawResponse
from .registry import _segments
from .security import TokenPurpose, sign_token, uuid7
from .urls import _fill, _scalar
if TYPE_CHECKING:
    from .service import Service, _Snapshot


class _NoCookies(DefaultCookiePolicy):
    def set_ok(self, cookie: Any, request: Any) -> bool: return False


class ClientError(Exception):
    def __init__(self, status: int, message: str):
        super().__init__(message)
        self.status = status


@dataclass(frozen=True)
class ClientOperationSchemas:
    inputs: dict[str, dict[str, Any]]
    required_inputs: frozenset[str]
    output: dict[str, Any]


class ClientContract:
    """A dependency's exported BP contract; each field retains its own definitions."""
    def __init__(self, value: Any):
        self._schema = parse("BpSchemaOutputSchema", value)
        self._operations: dict[str, tuple[dict[str, Any], dict[str, Any], Any, dict[str, Any], Any]] = {}
        for view in self._schema["manifest"]["views"]:
            for path in [view["path"], *view["pathVariants"]]: _segments(path)
            for operation in view["operations"]:
                identifier = operation["operationId"]
                if identifier in self._operations: raise ValueError("Duplicate dependency operation ID")
                fields = {"params": view["paramsSchema"], "query": operation["querySchema"], "headers": operation["headersSchema"]}
                body = operation["bodySchema"]
                if not body:
                    body = document("JsonObjectSchema")
                    body["root"] = {"kind": "optional", "inner": body["root"]}
                fields["body"] = body
                schemas, keys = {}, {}
                for name, field in fields.items():
                    field = deepcopy(field if field else document("JsonObjectSchema"))
                    if "root" not in field: raise ValueError("Dependency schemas must be AnyVali documents")
                    if name != "body": field["root"].setdefault("default", {})
                    schemas[name] = av.import_schema(object_document({name: field}, unknown_keys="reject"))
                    optional = document("JsonValueSchema")
                    # Each declared field schema owns value validation and coercion.
                    optional["root"] = {"kind": "optional", "inner": {"kind": "unknown"}}
                    keys[name] = optional
                output = av.import_schema(operation["jsonResponseSchema"]) if operation["jsonResponseSchema"] else None
                self._operations[identifier] = (view, operation, av.import_schema(object_document(keys, unknown_keys="reject")), schemas, output)

    @property
    def plugin_id(self) -> str: return self._schema["manifest"]["pluginId"]

    def schema(self) -> dict[str, Any]: return deepcopy(self._schema)

    def json_operations(self) -> dict[str, ClientOperationSchemas]:
        """Owned documents for native authoring; AnyVali determines input presence."""
        result = {}
        for identifier, (_, _, _, fields, output) in self._operations.items():
            if output is None: continue
            inputs = {}
            for name, field in fields.items():
                exported = deepcopy(av.export_schema(field))
                exported["root"] = exported["root"]["properties"][name]
                inputs[name] = exported
            result[identifier] = ClientOperationSchemas(inputs, frozenset(name for name, field in fields.items() if not field.safe_parse({}).success), deepcopy(av.export_schema(output)))
        return result

    def _prepare(self, identifier: str, values: Any) -> tuple[dict[str, Any], dict[str, Any], dict[str, Any], str, str, Any]:
        if identifier not in self._operations: raise ValueError("Unknown dependency operation")
        view, operation, keys, fields, output = self._operations[identifier]
        if output is None: raise ValueError("Dependency operation does not declare JSON output")
        source = keys.parse({} if values is None else values)
        parsed = {}
        for name, schema in fields.items():
            parsed.update(schema.parse({name: source[name]} if name in source else {}))
        if any(not isinstance(parsed.get(name), dict) for name in ("params", "query", "headers")):
            raise ValueError("Dependency params, query and headers must be objects")
        selected = next(((variant, path) for variant in [view["path"], *view["pathVariants"]] if (path := _fill(variant, parsed["params"])) is not None), None)
        if selected is None: raise ValueError("Dependency route parameters do not select a path")
        return view, operation, parsed, selected[1], selected[0], output


class ServiceClients:
    """One lazy connection pool per service; shutdown cancels its pending calls."""
    def __init__(self, service: Service):
        self._service = service
        self._http: httpx.AsyncClient | None = None
        self._pending: set[asyncio.Task[Any]] = set()
        self._closed = False

    def scope(self, tenant_id: str, app_id: str) -> RequestClients:
        """Background callers may use declared service-mode dependencies only."""
        return RequestClients(self, tenant_id, app_id)

    async def _send(self, client: Client, identifier: str, values: Any) -> Any:
        if self._closed: raise ClientError(503, "Dependency client is closed")
        task = asyncio.current_task(); assert task is not None
        self._pending.add(task)
        try:
            view, operation, parsed, path, variant, output = client._contract._prepare(identifier, values)
            state, scope = client._context._current()
            origin, headers = client._authorize(state, scope, view, operation, variant)
            if any(not isinstance(value, (str, int, float, bool)) for value in parsed["headers"].values()):
                raise ValueError("Dependency header values must be strings, numbers or booleans")
            custom = {name: _scalar(value) for name, value in parsed["headers"].items()}
            RawResponse(headers=custom)
            forbidden = {"authorization", "cookie", "set-cookie", "host", "origin", "referer", "accept", "accept-encoding", "content-type", "content-encoding"}
            if len({name.lower() for name in custom}) != len(custom) or any(name.lower() in forbidden or name.lower().startswith("x-bp-") for name in custom):
                raise ValueError("Dependency headers cannot replace framework credentials or routing")
            headers.update(custom)
            pairs = []
            for name, value in parsed["query"].items():
                if value is None: continue
                for item in value if isinstance(value, list) else [value]:
                    if item is None: continue
                    pairs.append((name, _scalar(item)))
            address = origin + path + ("?" + urlencode(pairs) if pairs else "")
            data = json.dumps(parsed["body"], ensure_ascii=False, separators=(",", ":"), allow_nan=False).encode() if "body" in parsed else None
            if len(address) > 8192 or data is not None and len(data) > 16 * 1024 * 1024:
                raise ValueError("Dependency request exceeds its size limit")
            if data is not None and operation["method"] in ("GET", "HEAD"):
                raise ValueError("Dependency GET/HEAD requests cannot carry a body")
            if data is not None: headers["content-type"] = "application/json"
            headers["accept"] = "application/json"
            headers["accept-encoding"] = "identity"
            if self._http is None:
                self._http = httpx.AsyncClient(follow_redirects=False, trust_env=False, timeout=30, cookies=CookieJar(policy=_NoCookies()))
            async def request():
                async with self._http.stream(operation["method"], address, content=data, headers=headers) as response:
                    if not 200 <= response.status_code < 300:
                        raise ClientError(response.status_code, "Dependency returned an unsuccessful status")
                    if response.status_code not in (204, 205) and response.headers.get("content-type", "").split(";", 1)[0].strip().lower() != "application/json":
                        raise ClientError(502, "Dependency did not return JSON")
                    if response.headers.get("content-encoding", "identity").strip().lower() != "identity":
                        raise ClientError(502, "Dependency returned unsupported content encoding")
                    body = bytearray()
                    async for chunk in response.aiter_bytes(chunk_size=8192):
                        if len(body) + len(chunk) > 16 * 1024 * 1024: raise ClientError(502, "Dependency response exceeds its size limit")
                        body.extend(chunk)
                    client._context._current(expected=state)
                    return output.parse(None if response.status_code in (204, 205) and not body else loads(body.decode("utf-8")))
            try: return await asyncio.wait_for(request(), 30)
            except ClientError: raise
            except Exception: raise ClientError(502, "Dependency request or response validation failed") from None
        finally: self._pending.discard(task)

    async def aclose(self) -> None:
        self._closed = True
        tasks = tuple(self._pending - {asyncio.current_task()})
        for task in tasks: task.cancel()
        await asyncio.gather(*tasks, return_exceptions=True)
        if self._http is not None: await self._http.aclose()


class RequestClients:
    """Opaque request binding; bearer credentials never enter render contexts."""
    def __init__(self, owner: ServiceClients, tenant_id: str, app_id: str, *, revision: _Snapshot | None = None, user_token: str | None = None):
        self._owner, self._tenant, self._app = owner, tenant_id, app_id
        self._revision, self._user_token = revision, user_token

    def _current(self, expected: _Snapshot | None = None) -> tuple[_Snapshot, ScopedContext]:
        service = self._owner._service
        state = service._state
        if not service.ready or state is None or self._revision is not None and state is not self._revision or expected is not None and state is not expected:
            raise ClientError(503, "Dependency configuration is unavailable or changed")
        if service._tenant_lock is not None and self._tenant != service._tenant_lock:
            raise ClientError(403, "Dependency tenant is not allowed")
        scope = state.snapshot.by_id(self._tenant, self._app)
        if scope is None: raise ClientError(403, "Dependency tenant/app is unavailable")
        return state, scope

    def user(self, contract: ClientContract, service_id: str | None = None) -> Client:
        if self._revision is None: raise ValueError("User clients require a service request context")
        return Client(self, contract, service_id=service_id or contract.plugin_id)

    def m2m(self, request_id: str, contract: ClientContract) -> Client:
        return Client(self, contract, request_id=request_id)


class Client:
    def __init__(self, context: RequestClients, contract: ClientContract, *, service_id: str | None = None, request_id: str | None = None):
        self._context, self._contract = context, contract
        self._service_id, self._request_id = service_id, request_id

    async def request(self, operation_id: str, values: Any = None) -> Any:
        return await self._context._owner._send(self, operation_id, values)

    def _authorize(self, state: _Snapshot, scope: ScopedContext, view: dict[str, Any], operation: dict[str, Any], path: str) -> tuple[str, dict[str, str]]:
        service = self._context._owner._service
        if self._request_id is None:
            assert self._service_id is not None
            reference = service.registry.dependencies.get(self._service_id, self._service_id)
            targets = [item for item in scope.tenant["services"] if item["enabled"] and reference in (item["id"], item.get("serviceId")) and item.get("serviceId") == self._contract.plugin_id]
            if len(targets) != 1: raise ClientError(403, "Dependency service is unavailable or ambiguous")
            target = targets[0]
            if "user" not in operation["auth"]["callers"]: raise ClientError(403, "Dependency operation does not allow user calls")
            app = {**scope.app, "routes": scope.app.get("appRoutes", scope.app["routes"])}
            access = AppAccess(ScopedContext(scope.tenant, app, scope.origin_policy), [target["id"]])
            if not access.allows_operation(view["viewId"], operation["operationId"], operation["method"], path, service_id=target["id"]):
                raise ClientError(403, "Dependency operation is not mounted")
            origins = [origin for hostname in scope.app["hostnames"] for origin in _origins(hostname)]
            if not origins: raise ClientError(503, "Dependency app origin is unavailable")
            headers = {"origin": origins[0]}
            if self._context._user_token: headers["authorization"] = "Bearer " + self._context._user_token
            elif operation["auth"]["required"]: raise ClientError(401, "Dependency requires a BP user")
        else:
            requests = [item for item in service.manifest["m2mRequests"] if item["id"] == self._request_id]
            if len(requests) != 1: raise ClientError(403, "Dependency request is not uniquely declared")
            request = requests[0]; mode = request["mode"]
            if mode not in operation["auth"]["callers"] or operation["method"] not in request["methods"]:
                raise ClientError(403, "Dependency caller mode or method is not allowed")
            policy = state.config.get("m2m", {})
            bindings = [item for item in policy.get("bindings", []) if item["enabled"] and item["requestId"] == self._request_id
                and item["sourceServiceId"] in policy["localServiceIds"] and item["tenantId"] == scope.tenant_id
                and item.get("appId", scope.app_id) == scope.app_id and item["mode"] == mode
                and item["contractId"] == request["contractId"] and item["targetViewId"] == view["viewId"]]
            if len(bindings) != 1: raise ClientError(403, "Dependency binding is unavailable or ambiguous")
            binding = bindings[0]
            contracts = [item for item in self._contract._schema["manifest"]["apiContracts"] if item["id"] == binding["contractId"] and item["viewId"] == view["viewId"]
                         and operation["method"] in item["methods"] and mode in item["modes"]]
            if len(contracts) != 1 or binding["contractId"] not in [item["id"] for item in operation["apiContracts"]]:
                raise ClientError(403, "Dependency contract does not cover the operation")
            if not set(request["requiredCapabilities"]).issubset(contracts[0]["capabilities"]):
                raise ClientError(403, "Dependency contract capabilities are insufficient")
            required = {*request["permissions"], *contracts[0]["permissions"], *(permission for item in operation["auth"]["permissions"] for permission in item["permissions"])}
            grants = [item for item in policy.get("grants", []) if item["enabled"] and item["bindingId"] == binding["id"] and item["tenantId"] == scope.tenant_id
                and item.get("appId", scope.app_id) == scope.app_id and operation["method"] in item["methods"] and required.issubset(item["permissions"])]
            if not grants: raise ClientError(403, "Dependency grant is unavailable or insufficient")
            targets = [item for item in policy["services"] if item["id"] == binding["targetServiceId"] and item["serviceId"] == self._contract.plugin_id]
            if len(targets) != 1: raise ClientError(403, "Dependency target is unavailable or ambiguous")
            target = targets[0]
            key, identity = service._signing_key, state.config.get("serviceIdentity", {})
            if key is None or identity.get("keyId") != key.kid or "".join(identity.get("publicKeyPem", "").split()) != "".join(key.public_key_pem.split()):
                raise ClientError(503, "Dependency signing identity is not registered")
            headers = {"x-bp-service-id": binding["sourceServiceId"], "x-bp-tenant-id": scope.tenant_id, "x-bp-app-id": scope.app_id}
            now = int(time.time())
            token = sign_token(key, {"iss": binding["sourceServiceId"], "sub": binding["sourceServiceId"], "aud": target["id"],
                "tenantId": scope.tenant_id, "appId": scope.app_id, "bindingId": binding["id"], "iat": now, "exp": now + 60, "jti": uuid7(), "tokenType": "service"}, TokenPurpose.SERVICE)
            if mode == "delegated":
                if not self._context._user_token or self._context._revision is None: raise ClientError(401, "Delegated dependency calls require a BP user")
                headers.update(authorization="Bearer " + self._context._user_token)
                headers["x-bp-service-authorization"] = "Bearer " + token
            else: headers["authorization"] = "Bearer " + token
        endpoint = secure_endpoint(target["hostname"])
        return http_origin(endpoint, allow_path=True) + httpx.URL(endpoint).raw_path.decode("ascii").rstrip("/"), headers
